crates/core/tedge/src/cli/mqtt/publish.rs

```diff
@@ -135,6 +135,13 @@ fn publish(cmd: &MqttPublishCommand) -> Result<(), MqttError> {
     }
 
     client.disconnect()?;
+    for event in connection.iter() {
```

Comment:
Initially I found the placement of this loop here, after the disconnect call, very confusing. But, I'm assuming that the disconnection request is also not processed until we explicitly poll the event loop as per the rumqtt design. Never expected this for a disconnect call as I thought that the `Outgoing::Disconnect` event is received by the event loop **after** the disconnect message is already sent to the broker, just as an ack to the client which doesn't need to be actively polled. But I guess they went for consistency in the processing model for disconnect as well.

Reply:
This was the root cause of the issue. Triggering a `disconnect()` is not enough.

However, the point of polling the rumqtt event loop for an `Outgoing::Disconnect` event  is:
- not to perform the disconnection (this is done by rumqtt in a background thread)
- but to be sure the disconnection actually happened before the `tedge mqtt pub` command dies.

Reply:
> but to be sure the disconnection actually happened before the tedge mqtt pub command dies.

In that case, should we wait for the `Event::Incoming(Incoming::Disconnect(_))` ack from the broker as well? To make sure that the broker received it as well?

Reply:
> In that case, should we wait for the `Event::Incoming(Incoming::Disconnect(_))` ack from the broker as well? To make sure that the broker received it as well?

No, this event is sent when the broker itself triggers the disconnection.